### PR TITLE
Feature/houston scraper cont

### PR DESCRIPTION
**Pull request by Shak2000**
### Link to Relevant Issue

This pull request resolves #106 

### Description of Changes

I refactored the Houston scraper

### CODE REVIEW

**Comment by dphoria** [comment]
Thank you as always, Shak. I think we should first fix the pipeline; which I have already done in a different PR: #122 . Feel free to merge that PR branch here, or save/apply the diff, whatever. Then we should be able to get the tests passing (at least the linting, etc.).

**Comment by dphoria** [comment]
> Thank you as always, Shak. I think we should first fix the pipeline; which I have already done in a different PR: #122 . Feel free to merge that PR branch here, or save/apply the diff, whatever. Then we should be able to get the tests passing (at least the linting, etc.).

Update: that PR was just merged in, Shak. So if you pull `main` into your PR branch, linting etc. should now pass.

**Comment by Shak2000** [comment]
> > Thank you as always, Shak. I think we should first fix the pipeline; which I have already done in a different PR: #122 . Feel free to merge that PR branch here, or save/apply the diff, whatever. Then we should be able to get the tests passing (at least the linting, etc.).
> 
> Update: that PR was just merged in, Shak. So if you pull `main` into your PR branch, linting etc. should now pass.

Thanks @dphoria! I merged the changes. There are a couple of tests in `test_scrapers.py` that can help you see the results for my PR.

File: cdp_scrapers/instances/houston.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+        """
+        log.info("start get body name")
+        event = self.remove_extra_type(event)
+        bodyTable = event.find_all("table")[1].find("table")
+        if "CITY COUNCIL" in bodyTable.text:
+            return "City Council"
+        else:
+            return bodyTable.find_all("span")[3].text.title()
+
+    def get_event_Minutes_Item(
```
Comment (dphoria) [inline cdp_scrapers/instances/houston.py:59]: Prefer consistent casing; the letter `m` should be in lower case.

File: cdp_scrapers/instances/houston.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+        super().__init__(timezone="America/Chicago")
+
+    def remove_extra_type(self, element: Union[Tag, NavigableString, None]) -> Tag:
+        """
+        Remove types that are not useful
+
+        Parameter:
+        ----------------
+        event: Union[Tag, NavigableString, None]
+            All elements in the page that we want to scrape
```
Comment (dphoria) [inline cdp_scrapers/instances/houston.py:26]: I don't really agree with this docstring. The word `element` implies a single object, not "all elements...". This just looks like some copy-paste job that was forgotten to be fixed later. There are other docstrings below with the same issue.

File: cdp_scrapers/instances/houston.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+            if name is not None and name != "":
+                event_minutes_items.append(
+                    ingestion_models.EventMinutesItem(
+                        minutes_item=ingestion_models.MinutesItem(name.strip())
+                    )
+                )
+
+        return event_minutes_items
+
+    # Big Functions
```
Comment (dphoria) [inline cdp_scrapers/instances/houston.py:94]: I think we should be consistent and stick to the numpy style docstrings that we use. I'd like to remove this rather random and meaningless comment.

File: cdp_scrapers/instances/houston.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    def get_diff_yearid(self, time: datetime) -> str:
```
Comment (dphoria) [inline cdp_scrapers/instances/houston.py:95]: `time` can be confusing, e.g. `datetime.time` or `time` modules. Would be better to use a less common word for this variable name.

File: cdp_scrapers/instances/houston.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+
+        Parameters:
+        --------------
+        time: datetime
+            The date of one event
+
+        Returns:
+        --------------
+        str
+            The main link, make agenda and video url in other function
```
Comment (dphoria) [inline cdp_scrapers/instances/houston.py:127]: "make agenda and video url in other function" seems to be irrelevant for this particular function. If anything, I _think_ that text blob should be in a different docstring section, e.g. [Notes](https://numpydoc.readthedocs.io/en/latest/format.html#notes).

File: cdp_scrapers/instances/houston.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+
+    def get_event(self, element_list) -> ingestion_models.EventIngestionModel:
+        """
+        Parse one event at a specific date. City council meeting information for
+        a specific date
+
+        Parameters:
+        --------------
+        event_time: datetime
+            Meeting date
```
Comment (dphoria) [inline cdp_scrapers/instances/houston.py:162]: The function signature uses `element_list` as a parameter, but the docstring uses `event_time`. Also, typing is missing for the parameter in the function signature.

File: cdp_scrapers/instances/houston.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+        --------------
+        List[BeautifulSoup]
+            Elements that contain different meetings
+        """
+        if time_from.year != time_to.year:
+            raise ValueError(
+                "time_from and time_to are in different years, which is not supported"
+            )
+        elements = []
+        main_URL = "https://houstontx.new.swagit.com/views/408"
```
Comment (dphoria) [inline cdp_scrapers/instances/houston.py:210]: For consistency, prefer `main_url`

File: cdp_scrapers/instances/atlanta.py [hunk 8]
```diff
@@ -774,7 +774,7 @@ def get_date(
     Returns:
     --------------
     list
-        all the ingestion models for the selected date range
+        all the ingestion models for the selected date ranges
```
[the remark this answers is not in the document]
Comment (dphoria) [reply cdp_scrapers/instances/atlanta.py:777]: Also, to be honest, I think we should exclude modifications to this file `atlanta.py`. It is not related to this PR, in my opinion.

**Review by dphoria — CHANGES_REQUESTED** [review]
Thank you very much for your work Shak. I have several comments, but mostly on styling.

File: cdp_scrapers/instances/houston.py (hunk 1 above)
Comment (Shak2000) [reply cdp_scrapers/instances/houston.py:59]: Done

File: cdp_scrapers/instances/houston.py (hunk 2 above)
Comment (Shak2000) [reply cdp_scrapers/instances/houston.py:26]: Done

File: cdp_scrapers/instances/houston.py (hunk 3 above)
Comment (Shak2000) [reply cdp_scrapers/instances/houston.py:94]: Done

File: cdp_scrapers/instances/houston.py (hunk 4 above)
Comment (Shak2000) [reply cdp_scrapers/instances/houston.py:95]: Done

File: cdp_scrapers/instances/houston.py (hunk 5 above)
Comment (Shak2000) [reply cdp_scrapers/instances/houston.py:127]: Done

File: cdp_scrapers/instances/houston.py (hunk 6 above)
Comment (Shak2000) [reply cdp_scrapers/instances/houston.py:162]: Done

File: cdp_scrapers/instances/houston.py (hunk 7 above)
Comment (Shak2000) [reply cdp_scrapers/instances/houston.py:210]: Done

File: cdp_scrapers/instances/houston.py (hunk 6 above)
Comment (dphoria) [reply cdp_scrapers/instances/houston.py]: One more nitpick Shak 😅 currently we have `element_list: []`. Personally not too thrilled about that. `[]` is an object, not a type. i.e. To me, it would be better to do something like `element_list: List`.

Actually, in fact, that's not the correct type at all. If you look at [how `get_event()` is called](https://github.com/CouncilDataProject/cdp-scrapers/blob/2b5b3700c6082ca8c1c3ed49bc85931ed5a2af0b/cdp_scrapers/instances/houston.py#L248-L249), it's called with a _single_ "thing" as the parameter.

```Python
for element in elements:
    events.append(self.get_event(element))
```

Comment (dphoria) [reply cdp_scrapers/instances/houston.py]: Well, I guess `element` there _could_ be a list of some kind. But if so, then that is very misleading. So, either just change the typing in the function signature from `[]` to something like `List` (nitpicky LOL); or, change the variable name at the call site (in `get_events()`) so that it's clear it's a list (`element` just sounds like "one thing").

Comment (Shak2000) [reply cdp_scrapers/instances/houston.py]: Good catch! I think I had a mixing of previous names and current implementation.

In this function `get_all_elements_in_range`, I need to return both the date and the element, so I am using a dictionary.

File: cdp_scrapers/instances/houston.py [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+        Returns:
+        --------------
+        list[ingestion_models.EventIngestionModel]
+            A list of EventIngestionModel that contains all city council
+            meetings information within a specific time range
+        """
+        events = []
+        d = self.get_all_elements_in_range(from_dt, to_dt)
+        for date in d.keys():
+            events.append(self.get_event(date, d[date]))
```
Comment (dphoria) [inline cdp_scrapers/instances/houston.py:256]: More pythonic to do something like this:

```Python
for date, element in d.items():
    events.append(self.get_event(date, element))
```

Also _slightly_ more efficient since this avoids additional hashing/hash table lookup.

File: cdp_scrapers/instances/portland.py [hunk 10]
```diff
@@ -179,19 +179,24 @@ def separate_name_from_title(title_and_name: str) -> str:
...
+
+    while (
+        "Former" in title_and_name
+        or "Mayor" in title_and_name
+        or "Commissioner" in title_and_name
+    ):
+        name_index = title_and_name.find(" ")
+        title_and_name = title_and_name[name_index + 1 :]
+
+    return title_and_name.strip()
```
Comment (dphoria) [inline cdp_scrapers/instances/portland.py:199]: Shak, are these changes to `portland.py` necessary to get the pipeline pass? If not, personally, I think we should exclude this, and merge via a separate PR. I prefer to keep PR scope small, if possible. :sweat_smile:

**Review by dphoria — CHANGES_REQUESTED** [review]
Thanks again for your patience, Shak. Just a few more comments/requests. :pray:

File: cdp_scrapers/instances/houston.py [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+        Returns:
+        --------------
+        str
+            The main link for this event
+        """
+        link_post = element.find("a")["href"]
+        link = f"https://houstontx.new.swagit.com/{link_post}"
+        return link
+
+    def get_agenda(self, element: BeautifulSoup) -> Union[Tag, NavigableString, None]:
```
[the remark this answers is not in the document]
Comment (Shak2000) [reply cdp_scrapers/instances/houston.py:136]: Thanks! Done

File: cdp_scrapers/instances/houston.py (hunk 9 above)
Comment (Shak2000) [reply cdp_scrapers/instances/houston.py:256]: Thanks! Done

File: cdp_scrapers/instances/portland.py (hunk 10 above)
Comment (Shak2000) [reply cdp_scrapers/instances/portland.py:199]: I made these changes to `portland.py` because the tests on GitHub failed, and I needed to ensure that the scraper would pass them. See [this discussion on Slack](https://app.slack.com/client/TMPTLMNJ2/CMA9K2F6E/thread/CMA9K2F6E-1671340475.739399) for more details.

File: cdp_scrapers/instances/houston.py [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+        return year_id
+
+    def get_date_mainlink(self, element: Tag) -> str:
+        """
+        Find the main link for one event.
+
+        Parameters:
+        --------------
+        time: datetime
+            The date of one event
```
Comment (dphoria) [inline cdp_scrapers/instances/houston.py:125]: This is still incorrect. This function takes `element` as a parameter, not `time`. Same problem for the docstring in `get_agenda()` below.

**Review by dphoria — CHANGES_REQUESTED** [review]
I think will be it, Shak. One more comment on old docstrings.

File: cdp_scrapers/instances/houston.py (hunk 12 above)
Comment (Shak2000) [reply cdp_scrapers/instances/houston.py:125]: Done

**Review by dphoria — APPROVED** [review]

**Comment by dphoria** [comment]
Thank you Shak, especially taking care of all my nitpicky review comments.

**Comment by evamaxfield** [comment]
/test-scraper get_houston_events --from_dt=2022-10-14T00:00:00 --to_dt=2022-10-22T00:00:00
